src/QuickCheckVEngine/Templates/Utils/HPM.hs

```diff
@@ -137,5 +138,13 @@ surroundWithHPMAccess_core shrink evt x tmpReg = Random $ do
                  <> triggerHPMCounter tmpReg hpmCntIdx
   let epilogue = uniformTemplate [ readHPMCounter  tmpReg hpmCntIdx
                                  , readHPMCounterM tmpReg hpmCntIdx ]
-  return $ if shrink then prologue <> x <> epilogue
-                     else NoShrink prologue <> x <> NoShrink epilogue
+  --let prolo = if countInst then prologue <> csrr instReg2 0xB02 else prologue
+  --let epilo = if countInst then csrr instReg3 0xB02 <> epilogue else epilogue
+  return $ if shrink then prologue <> surroundWithHPMAccess_raw x instRet <> epilogue
+                     else NoShrink prologue <> surroundWithHPMAccess_raw x instRet <> NoShrink epilogue
+
+
+surroundWithHPMAccess_raw :: Template -> Maybe (Integer, Integer) -> Template
+surroundWithHPMAccess_raw x instRet = case (instRet) of
```

Comment:
I think I'd use the pattern matching for function definition syntax here... or possibly the case match syntax as you have it, but with a 2 spaces indentation for the case alternatives